src/v2/models/AppState.js

```diff
@@ -241,33 +243,47 @@ export default Model.extend({
     this.trigger('cache:clear');
   },
 
-  setIonResponse(transformedResponse) {
+  async setIonResponse(transformedResponse, hooks) {
     if (!this.shouldReRenderView(transformedResponse)) {
       return;
     }
 
-    // `currentFormName` is default to first form of remediations or nothing.
-    let currentFormName = null;
-    if (!_.isEmpty(transformedResponse.remediations)) {
-      currentFormName = transformedResponse.remediations[0].name;
-    } else {
-      Logger.error('Panic!!');
-      Logger.error('\tNo remediation found.');
-      Logger.error('\tHere is the entire response');
-      Logger.error(JSON.stringify(transformedResponse, null, 2));
-    }
+    const currentFormName = this.getFormName(transformedResponse);
 
     this.clearAppStateCache();
 
     // set new app state properties
     this.set(transformedResponse);
 
+    const hook = hooks?.getHook(currentFormName); // may be undefined
+    await executeHooksBefore(hook);
+
     // make sure change `currentFormName` is last step.
     // change `currentFormName` will re-render FormController,
     // which may depend on other derived properties hence
     // those derived properties must be re-computed before
     // re-rendering controller.
     this.set({ currentFormName });
-  }
+
+    await executeHooksAfter(hook);
```

Comment:
should we add a check that execute only if there is any hook exists..

Reply:
I have included a unit test to verify that passing an undefined hook will not error, and will be a no-op.